assets/js/modules/analytics-4/datastore/partial-data.js

```diff
@@ -0,0 +1,578 @@
+				return true;
+			}
+
+			const { startDate } = select( CORE_USER ).getDateRangeDates( {
+				offsetDays: DATE_RANGE_OFFSET,
+			} );
+
+			// TODO: use `replaceAll` instead when we upgrade our Node version.
+			const startDateYYYYMMDD = Number( startDate.replace( /-/g, '' ) );
+
+			return dataAvailabilityDate > startDateYYYYMMDD;
+		}
+	),
+
+	/**
+	 * Gets whether the given audience is in partial data state.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state        Data store's state.
+	 * @param {string} audienceSlug Audience slug.
+	 * @return {boolean} Returns TRUE if partial data, otherwise FALSE or undefined while loading.
+	 */
+	isAudiencePartialData: createRegistrySelector(
+		( select ) => ( state, audienceSlug ) =>
+			select( MODULES_ANALYTICS_4 ).isResourcePartialData(
+				audienceSlug,
+				RESOURCE_TYPE_AUDIENCE
+			)
+	),
+
+	/**
+	 * Gets whether the given custom dimension is in partial data state.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state               Data store's state.
+	 * @param {string} customDimensionSlug Custom dimension slug.
+	 * @return {boolean} Returns TRUE if partial data, otherwise FALSE or undefined while loading.
+	 */
+	isCustomDimensionPartialData: createRegistrySelector(
+		( select ) => ( state, customDimensionSlug ) =>
+			select( MODULES_ANALYTICS_4 ).isResourcePartialData(
+				customDimensionSlug,
+				RESOURCE_TYPE_CUSTOM_DIMENSION
+			)
+	),
+
+	/**
+	 * Gets whether the given property is in partial data state.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state      Data store's state.
+	 * @param {string} propertyID Property ID.
+	 * @return {boolean} Returns TRUE if partial data, otherwise FALSE or undefined while loading.
+	 */
+	isPropertyPartialData: createRegistrySelector(
+		( select ) => ( state, propertyID ) =>
+			select( MODULES_ANALYTICS_4 ).isResourcePartialData(
+				propertyID,
+				RESOURCE_TYPE_PROPERTY
+			)
+	),
+	/**
+	 * Gets report options for determining data availability date for a given resource.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state        Data store's state.
+	 * @param {string} resourceSlug Resource slug.
+	 * @param {string} resourceType Resource type.
+	 * @return {Object|undefined} Report options for determining data availability date. Undefined if not available.
+	 */
+	getPartialDataReportOptions: createRegistrySelector(
+		( select ) => ( state, resourceSlug, resourceType ) => {
+			const propertyID = select( MODULES_ANALYTICS_4 ).getPropertyID();
+
+			if ( ! propertyID ) {
+				return undefined;
+			}
+
+			const propertyCreateTime =
+				select( MODULES_ANALYTICS_4 ).getPropertyCreateTime();
+
+			if ( ! propertyCreateTime ) {
+				return undefined;
+			}
+
+			const endDate = select( CORE_USER ).getReferenceDate();
+
+			switch ( resourceType ) {
+				case RESOURCE_TYPE_AUDIENCE:
+					return {
+						startDate: getDateString(
+							new Date( propertyCreateTime )
+						),
```

Comment:
As this is duplicated across all types, how about we create a `startDate` variable above and just use that everywhere?